tests/migration_test.go

```diff
@@ -931,24 +932,59 @@ var _ = Describe("[rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][level:system
 		})
 		Context("with an Fedora shared NFS PVC, cloud init and service account", func() {
 			var pvName string
+			var nfsPodName string
 			var vmi *v1.VirtualMachineInstance
-			BeforeEach(func() {
+
+			initNFS := func(ipFamily k8sv1.IPFamily) {
 				tests.SkipNFSTestIfRunnigOnKindInfra()
 				pvName = "test-nfs" + rand.String(48)
 				// Prepare a NFS backed PV
 				By("Starting an NFS POD")
 				os := string(cd.ContainerDiskFedora)
-				nfsIP := tests.CreateNFSTargetPOD(os)
+
 				// create a new PV and PVC (PVs can't be reused)
 				By("create a new NFS PV and PVC")
+				nfsPod := tests.CreateNFSTargetPOD(os)
+				nfsPodName = nfsPod.Name
+				nfsIP := libnet.GetPodIpByFamily(nfsPod, ipFamily)
+				ExpectWithOffset(1, nfsIP).NotTo(BeEmpty())
 				tests.CreateNFSPvAndPvc(pvName, "5Gi", nfsIP, os)
+			}
+
+			BeforeEach(func() {
+				pvName = ""
+				nfsPodName = ""
+				vmi = nil
 			})
 
 			AfterEach(func() {
-				// PVs can't be reused
-				tests.DeletePvAndPvc(pvName)
+				if vmi != nil {
+					By("Deleting the VMI")
+					Expect(virtClient.VirtualMachineInstance(vmi.Namespace).Delete(vmi.Name, &metav1.DeleteOptions{})).To(Succeed())
+				}
 			})
-			It("[test_id:2653]  should be migrated successfully, using guest agent on VM", func() {
+
+			AfterEach(func() {
+				if nfsPodName != "" {
+					By("Deleting NFS pod")
+					Expect(virtClient.CoreV1().Pods(tests.NamespaceTestDefault).Delete(nfsPodName, &metav1.DeleteOptions{})).To(Succeed())
+				}
+			})
+
+			AfterEach(func() {
+				if pvName != "" {
+					// PVs can't be reused
+					By("Deleting PV and PVC")
+					tests.DeletePvAndPvc(pvName)
+				}
+			})
```

Comment:
Its kinda awkward to have more then one `AfterEach` block I am quite sure we dont do that in any test.
If I understand correctly you split `AfterEach` to three so if one object fails to delete others will be delete on another block, right?
If so I am not sure this is the right direction, 
the VM is the NFS PV and PVC consumer, so if the VM wont delete the PV and PVC wont delete as well and since `tests.DeletePVandPVC` uses panics if an error occurs I think `Context` block that contains this `AfterEach` will be skipped (correct me if I am wrong) so other objects wont delete.
I think this is why deleting the VMI and target pod were in the `It` block

Also the order of the teardown should be:
    delete VM and wait for it to dispose -> PVC -> PV -> NFS target
Deleting the NFS target pod before the PV might cause errors.
I think its better to keep single `AfterEach` block and delete the objects this test create in the order I mentioned.

At the current state of test suite there the function `tests.BeforeTestCleanup` used in most (if not all) tests `BeforeEach` block, this function deletes VM, PV's and PVC's as well
Until this method of "use cleanup before each test" will be replaced we shouldn't care for the teardown, what do I miss?

Reply:
> Its kinda weird to have more then one `AfterEach` block I am quite sure we dont do that in any test.
> If I understand correctly you split `AfterEach` to three so if one object fails to delete others will be delete on another block, right?
> If so I am not sure this is the right direction,

This pattern of multiple `AfterEach` blocks is used in the [services tests](https://github.com/kubevirt/kubevirt/blob/master/tests/network/services.go#L170) ; while it also struct me as weird at first, I agree that having multiple blocks is the better option, since (as you've correctly said), it covers scenarios where "if one object fails to delete others will be delete on another block".

Reply:
AFAIU the k8s pattern. The order of resource removal is not important. If there is a resource that other resources dependa on. It should have a finalizer preventing from it being removed till the depending resource is removed.

Anyway, I returned back the waiting for removes since it is not the focus of this PR and I would prefer to discuss it outside of this PR.
Now let's hope the 'AfterEach` order is constant:)

Reply:
If you are not able to cleanup (e.g: a VM wont delete) then the cluster is no longer in a "clean state" for next tests, so you should stop test suite and raise an error.

Just a note, at the current state of tests suite we clean the cluster before a test starts (using `tests.CleanupBeforeTest` at `BeforeEach` block in any context) I guess so we could have artifacts of all the objects a test created.
We lose those artifacts if we would delete those objects at AfterEach.

Reply:
We are not stopping the run of the test suite if a cleanup fails (we can open a separate discussion on whether we should). So it is better to clean what we can and not stop cleaning upon an error.